From c876dc534fe0677bc7341a4cda7b51fc71453b6c Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:04:37 -0400
Subject: [usb] hub: Use correct reset for wedged USB3 devices that are NOTATTACHED

Message-id: <1410887463-20674-206-git-send-email-dzickus@redhat.com>
Patchwork-id: 95215
O-Subject: [RHEL7 PATCH 205/591] usb: hub: Use correct reset for wedged USB3 devices that are NOTATTACHED
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 2d51f3cd11f414c56a87dc018196b85fd50b04a4
Author: Julius Werner <jwerner@chromium.org>
Date:   Thu Nov 7 10:59:14 2013 -0800

    usb: hub: Use correct reset for wedged USB3 devices that are NOTATTACHED

    This patch adds a check for USB_STATE_NOTATTACHED to the
    hub_port_warm_reset_required() workaround for ports that end up in
    Compliance Mode in hub_events() when trying to decide which reset
    function to use. Trying to call usb_reset_device() with a NOTATTACHED
    device will just fail and leave the port broken.

    Signed-off-by: Julius Werner <jwerner@chromium.org>
    Acked-by: Alan Stern <stern@rowland.harvard.edu>
    Cc: stable <stable@vger.kernel.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index bd6986e..18a5ac3 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -4832,8 +4832,9 @@ static void hub_events(void)
      hub->ports[i - 1]->child;
 
     dev_dbg(hub_dev, "warm reset port %d\n", i);
-    if (!udev || !(portstatus &
-      USB_PORT_STAT_CONNECTION)) {
+    if (!udev ||
+        !(portstatus & USB_PORT_STAT_CONNECTION) ||
+        udev->state == USB_STATE_NOTATTACHED) {
      status = hub_port_reset(hub, i,
        NULL, HUB_BH_RESET_TIME,
        true);
-- 
1.7.1